README.markdown
Roar - Python based, easy extendable, blog aware, static site generator
=======================================================================

Roar is a static website generator, forked from
[Growl](http://github.com/xfire/growl/tree), which in turn is heavily
inspired from [Jekyll](http://github.com/mojombo/jekyll/tree/master),
and which shamelessly stole some really cool ideas from Jekyll.

Nevertheless Growl brought some nice features:

* Minimal dependencies
* Easy to install (and use? ;])
* Easy to extend

The [Growl based site of xfire's blog](http://github.com/xfire/downgrade/tree)
is also available on github.

On top, Roar departs from Growl in some aspects:

* Even less dependencies (e.g. Yaml is no longer used)
* Packaged for the PyPi
* Actively maintained


Installation
------------

### Prequisites

The following basic packages are needed:

    > apt-get install python

All other is optional depending on you own needs.

I recommend using [Jinja2][jinja2] as the templating engine. Roar will
use [Jinja2][jinja2] as default, if it is installed.

    > apt-get install python-jinja2

You are free to use some other templating engine like [Django][django],
[Mako][mako] or [Cheetah][cheetah]. for examples how to
configure them, see [extending Roar](#extending_roar).

### Finish the installation

After installing all needed packages, you can use `roar`
directly or copy it to a place which is in your `$PATH`.

    > ./roar ...
    > cp roar /usr/local/bin


Usage
-----

Simply call `roar` with the source directory:

    > roar my.site

Roar will then generate the output in the directory `my.site/_deploy`.
if you want Roar to spit the generated stuff into another directory,
simply specify this director as second parameter.

    > roar my.site /tmp/my.site.output

### options

* `--serve[:port]` (default port: 8080)

  Generate the site to the deploy directory and then start a simple
  webserver. this is intended to be used for testing purposes only.

	> roar my.site --serve 1234

* `--deploy`

  Trigger deploy process. this does nothing per default, but you can
  add actions using hooks. (see `_hooks/deploy_rsync.py`)


Input data
----------

Roar will ignore all files and directories which starts with
a `.` or a `_`. (this can be changed via `Site.IGNORE`, see
[extending Roar](#extending_roar))

All files ending with `_` or a transformer extension (`Config.transformers`)
are processed as **pages**. in these cases, the ending will be striped from
the filename. E.g.

* `index.html_`  ->  `index.html`
* `atom.xml_`  ->  `atom.xml`
* `somefile.txt.markdown`  ->  `somefile.txt`

Some directories beginning with an `_` are special to Roar:

* `_deploy/` the default deploy directory
* `_layout/` contains your site layouts
* `_posts/` contains your posts
* `_hooks/` contains all your hooks (see [extending roar](#extending_roar))
* `_libs/` contains third party code (see [extending roar](#extending_roar))

All **pages** and **posts** must have an `rfc822` header. An empty line
separates the header from the content. Example:

    layout: post
    title: my post title
    category: spam, eggs

    <html>
       <!-- more content here -->
    </html>


If no headers are needed, and *empty* `rfc822` header section consisting
of an empty line before the actual content will work.

All data defined in this header will be attached to the corresponding object
and can be accessed in your template code. an example in [Jinja2][jinja2] may
look like

    <ul>
    {% for post in site.posts|reverse %}
        <li>{{ post.title }} - {{ post.date }}</li>
    {% endfor %}
    </ul>

in the context of your template, you have access to one or more of the following
objects.

### site

This holds the site wide informations.

* `site.now`

  Current date and time when you run roar. this is a Python
  [datetime](http://docs.python.org/library/datetime.html#datetime-objects) object.

    {{ site.now.year }}-{{site.now.month}}

* `site.posts`

  Chronological list of all posts.

    {% for post in site.posts|reverse|slice(8) %}
        {{ post.content }}
    {% endfor %}

* `site.unpublished_posts`

  Chronological list of all unpublished posts. e.g. all posts which set
  `publish` to `false`.

* `site.categories`

  Dictionary mapping category <-> posts.

    <ul>
    {% for cat in site.categories %}
        <li> <stong>{{ cat }}</strong>
            <ul>
                {% for post in site.categories.get(cat) %}
                    <li><a href="{{ post.url }}">{{ post.title }}</a> - {{ post.date }}</li>
                {% endfor %}
            </ul>
        </li>
    {% endfor %}
    </ul>

### page

* `page.url`

  The relative URL to the page.

* `page.transformed`

  The transformed content. no layouts are applied here.

### post

* `post.date`

  A `datetime` object with the publish date of the post.

* `post.url`

  Relative URL to the post.

* `post.publish`

  If set to `false`, the post will be generated, but is not in the list
  of `site.posts`. Instead it's in the `site.unpublished_posts` list.

  If `publish` is not set, Roar will assume this as `true` and the post
  will be normally published.

* `post.content`

  The transformed content. Exactly the layout specified in the `rfc822`
  header is applied  (no recursive applying).

* `post.transformed`

  The transformed content. No layouts are applied here.


<a name="extending_roar"></a>
Extending Roar
--------------

Roar is very easy extendable via Python code placed in the `_hooks` and
`_libs` directory.

If the `_libs` directory exists, it is added to the Python module search path
(`sys.path`), so Python modules dropped there will be available in the code.

All files in the `_hooks` directory, which end with `.py`, are executed
directly in the global scope of the `roar` script. Thus a hook can freely
shape Roar's code at will. Roar loves that! ;)

Here are some examples of what can be done. But you sure can imagine other
cool things.


### Configuring template engines



### Register new transformers

New transformers can be registered in the `Config` class by adding a
filename extension <-> transformation function mapping to the `transformers`
attribute. here's an example for `markdown2`:

    import markdown2
    import functools

    Config.transformers['noop'] = lambda source: source
    Config.transformers['markdown2'] = functools.partial(
                markdown2.markdown,
                extras={'code-color': {"noclasses": True}})

The transformation function must return the transformed source text which is given
as the only parameter. so if you need to add more parameters to your
transformation function, best use the [functools](http://docs.python.org/library/functools.html)
module as you see in the example above.



### Change which files will be ignored

Roar decides to ignore files which filenames start with one of the tokens
defined in `Site.IGNORE`. so a hook with the following content will make
Roar ignore all files begining with `.`, `_` and `foo`.

    Site.IGNORE += ('foo',)



### Define global template context content

Simply add your content to `Site.CONTEXT` like these examples:

    Site.CONTEXT.author = 'Rico Schiekel'
    Site.CONTEXT.site = AttrDict(author = 'Rico Schiekel')

Note: `Site.CONTEXT.site` has to be an `AttrDict` instance!



### Add some verbosity

As an example, we would display the currently processed post, while
Roar chomps your input.

Create a new file (e.g. `verbose.py`) in the `_hooks` directory with the
following content:

    @wrap(Post.write)
    def verbose_post_write(forig, self):
        print 'post: %s - %s\n' % (self.date.strftime('%Y-%m-%d'), self.title)
        return forig(self)

Roar offers the helper decorator `wrap`, which wraps an existing method
of a class with a new one. the original method is given to your new function
as the first parameter (`forig`).




Bug reporting
-------------

Please report bugs [here](http://github.com/Connectical/roar/issues).


License
-------
[GPLv2](http://www.gnu.org/licenses/gpl-2.0.html)



  [jinja2]:  http://jinja.pocoo.org/2/          "jinja2"
  [django]:  http://www.djangoproject.com/      "django"
  [mako]:    http://www.makotemplates.org/      "mako"
  [cheetah]: http://www.cheetahtemplate.org/    "cheetah"



debian/changelog
roar (0.1.2-1) testing; urgency=low

  * Initial release.

 -- Oscar Garcia Amor <ogarcia@connectical.com>  Wed, 26 Oct 2011 10:00:44 +0200


_hooks/posts_ui.py
# vim:syntax=python:sw=4:ts=4:expandtab
#
# Copyright (C) 2009 Rico Schiekel (fire at downgra dot de)
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License version 2
# as published by the Free Software Foundation
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
# MA 02110-1301, USA.
#

import os
import time
import datetime
import tempfile
import textwrap
import readline
import urllib
import hashlib


def setup_posts_ui():

    def get_editor():
        editor = os.environ.get('ROAR_EDITOR')
        if not editor:
            editor = os.environ.get('EDITOR')
        if not editor:
            editor = 'vi'
        return editor

    def launch_editor(content = ''):
        fn = None
        try:
            fid, fn = tempfile.mkstemp('.post', 'roar_', None, True)
            f = open(fn, 'w')
            f.write(content)
            f.close()

            mod_time_start = os.stat(fn).st_mtime
            rcode = subprocess.call([get_editor(), fn])
            mod_time_end = os.stat(fn).st_mtime

            hash_org = hashlib.sha256(content).digest()
            f = open(fn, 'r')
            content = f.read()
            f.close()
            hash_new = hashlib.sha256(content).digest()

            if (rcode != 0 or
                    mod_time_end == mod_time_start or
                    hash_org == hash_new):
                return None

            # only delete temp file if anything went ok
            return content
        finally:
            if fn:
                os.unlink(fn)

    def raw_input_default(prompt, value = None):
        if value:
            readline.set_startup_hook(lambda: readline.insert_text(value))
        try:
            return raw_input(prompt)
        finally:
            if value:
                readline.set_startup_hook(None)

    def mangle_url(url):
        ou = url
        url = url.lower()
        url = ''.join(c for c in url if c not in mangle_url.SP)
        url = url.replace('&', ' and ')
        url = url.replace('.', ' dot ')
        url = url.strip()
        url = url.replace(' ', '_')
        return urllib.quote(url)
    mangle_url.SP = '`~!@#$%^*()+={}[]|\\;:\'",<>/?'

    def create_new_post(self):
        TEMPLATE = textwrap.dedent("""
        ---
        layout: post
        title: ???
        categories: ???
        ---
        """).strip()
        try:
            content = launch_editor(TEMPLATE)
            if content:
                # load yaml header
                mo = Template.RE_YAML.match(content)
                if mo and mo.groupdict().get('yaml'):
                    meta = yaml.load(mo.groupdict().get('yaml'))
                    title = meta.get('title')

                if title:
                    title = mangle_url(title)

                    tnow = datetime.datetime.now().timetuple()
                    filename = time.strftime('%Y-%m-%d-', tnow)
                    filename += title

                    try:
                        filename = raw_input_default('filename: ', filename)
                        filename = os.path.join(self.POST_DIR, filename)
                        f = open(filename, 'w')
                        f.write(content)
                        f.close()
                        print 'created post: %s' % filename
                    except KeyboardInterrupt:
                        # save backup to temp file
                        print '\nabort...'
                        fid, fn = tempfile.mkstemp('.post', 'roar_',
                                                   None, True)
                        f = open(fn, 'w')
                        f.write(content)
                        f.close()
                else:
                    print 'abort... (no title)'
            else:
                print 'abort...'
        except Exception, e:
            print 'can\'t create new post: %s' % e
            raise

    @wrap(Site.setupOptions)
    def setupOptions(forig, self, parser):
        forig(self, parser)
        parser.add_option('-n', '--newpost',
                          action = 'store_true', dest = 'new_post',
                          help = 'create new post')

    @wrap(Site.run)
    def site_run(forig, self):
        """ write all posts to the deploy directory.
        """
        if self.options.new_post:
            create_new_post(self)
        else:
            forig(self)

setup_posts_ui()


growl.py
#!/usr/bin/env python
#
# vim:syntax=python:sw=4:ts=4:expandtab
#
# Copyright (C) 2009 Rico Schiekel (fire at downgra dot de)
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License version 2
# as published by the Free Software Foundation
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
# MA 02110-1301, USA.
#

__author__ = 'Rico Schiekel <fire@downgra.de>'
__copyright__ = 'Copyright (C) 2009 Rico Schiekel'
__license__ = 'GPLv2'
__version__ = '0.3'


import os
import sys
import re
import shutil
import datetime
import collections
import itertools
import functools
import inspect
import codecs
import rfc822
from optparse import OptionParser


def renderTemplate(template, context):
    raise NotImplementedError('no template engine configured!')


try:
    import jinja2

    jinja2_env = jinja2.Environment()

    def renderTemplate(template, context):
        return jinja2_env.from_string(template).render(context)

    def templateFilter(func):
        """ decorator to easily create jinja2 filters
        """
        jinja2_env.filters[func.__name__] = func
except ImportError:
    pass


def wrap(orig_func):
    """ decorator to wrap an existing method of a class.
        e.g.

        @wrap(Post.write)
        def verbose_write(forig, self):
            print 'generating post: %s (from: %s)' % (self.title,
                                                      self.filename)
            return forig(self)

        the first parameter of the new function is the the original,
        overwritten function ('forig').
    """

    # har, some funky python magic NOW!

    def outer(new_func):

        @functools.wraps(orig_func)
        def wrapper(*args, **kwargs):
            return new_func(orig_func, *args, **kwargs)

        if inspect.ismethod(orig_func):
            setattr(orig_func.im_class, orig_func.__name__, wrapper)
        return wrapper
    return outer


class AttrDict(dict):
    """ dictionary which provides its items as attributes.
    """

    def __getattr__(self, name):
        return self[name]

    def __setattr__(self, name, value):
        self[name] = value

    def copy(self):
        return AttrDict(super(AttrDict, self).copy())


class Config(object):
    """ base class providing some static configuration values.
    """

    LIB_DIR = HOOK_DIR = ''

    @classmethod
    def updateconfig(cls, base, deploy):
        cls.transformers = {}
        cls.BASE_DIR = base
        cls.DEPLOY_DIR = deploy
        cls.LAYOUT_DIR = os.path.join(base, '_layout')
        cls.HOOK_DIR = os.path.join(base, '_hooks')
        cls.LIB_DIR = os.path.join(base, '_libs')
        cls.POST_FILE_EXT = '.html'


class Template(Config):
    """ abstract template base class providing support for an
        rfc822 header, transforming based on file extension,
        rendering (only using current layout if defined) and
        layouting (applying all layouts).
    """

    def __init__(self, filename, layouts, context):
        super(Template, self).__init__()
        self.filename = filename
        self.layouts = layouts
        self.context = context.copy()
        self.context.layout = None
        self.read_message()

    def read_message(self):
        """ read MIME headers and remove the headers from content
        """
        infile = file(self.filename, "rU")
        msg = rfc822.Message(infile)
        encoding = msg.get("encoding", "utf-8")
        self.context.update((k, v.decode(encoding)) for k, v in msg.items())
        self._content = infile.read().decode(encoding)
        infile.close()

    def transform(self):
        """ do transformation based on filename extension.
            e.g. do markdown or textile transformations
        """
        ext = os.path.splitext(self.filename)[-1][1:]
        t = self.transformers.get(ext, lambda x: x)
        return t(self._content)

    def render(self):
        """ render content, so transforming and then
            apply current layout.
        """
        ctx = self.context.copy()
        ctx.content = renderTemplate(self.transform(), ctx)
        layout = self.layouts.get(ctx.layout)
        if layout:
            return renderTemplate(layout.content, ctx)
        else:
            return ctx.content

    def layout(self):
        """ layout content, so transforming and then applying
            all layouts.
        """
        ctx = self.context.copy()
        ctx.content = self.render()
        layout = self.layouts.get(ctx.layout)
        if layout:
            layout = self.layouts.get(layout.layout)

        while layout != None:
            ctx.content = renderTemplate(layout.content, ctx)
            layout = self.layouts.get(layout.layout)

        return ctx.content

    def write(self, path, content):
        """ write content to path in deploy directory.
        """
        fname = os.path.join(self.DEPLOY_DIR, path)
        dirname = os.path.dirname(fname)
        if not os.path.isdir(dirname):
            os.makedirs(dirname)
        f = file(fname, 'w')
        f.write(content.encode("utf8"))
        f.close()

    def __getattr__(self, name):
        if not name in self.context:
            raise AttributeError("'%s' object has no attribute '%s'" %
                                    (self.__class__.__name__, name))
        return self.context[name]

    @property
    def transformed(self):
        return self.transform()


class Layout(Template):
    """ a layout template from _layouts/ directory.
    """

    def __init__(self, filename, context):
        super(Layout, self).__init__(filename, {}, context)

        base = os.path.basename(filename)
        ext = os.path.splitext(base)
        self.name = ext[0]

    @property
    def layout(self):
        return self.context.get('layout')

    @property
    def content(self):
        return self.transform()


class Page(Template):
    """ a page template which should be transformed. e.g. files which
        filename ends with an '_' or an transformer file extension.
    """

    TRANSFORM = ('_', )

    def __init__(self, filename, layout, context):
        super(Page, self).__init__(filename, layout, context)

        self.context.page = self

    @property
    def url(self):
        return self.path.replace(os.path.sep, '/')
        
    @property
    def urlparts(self):
        return self.url.split("/")
    
    @property
    def root(self):
        return "../" * self.url.count("/")
        
    @property
    def path(self):
        path = os.path.abspath(self.filename)
        npath, ext = os.path.splitext(path)
        if self.filename[-1] in Page.TRANSFORM:
            path = path[:-1]
        elif ext and ext[1:] in self.transformers:
            path = npath
        path = path.replace(os.path.abspath(self.BASE_DIR), '', 1)
        return path.lstrip(os.path.sep)

    @property
    def content(self):
        return self.render()

    def write(self):
        return super(Page, self).write(self.path, self.layout())

    @staticmethod
    def transformable(filename):
        """ return true, if the file is transformable. that means the
            filename ends with a character from self.TRANSFORM or
            self.transformers.
        """
        ext = os.path.splitext(filename)[-1]
        return ((filename[-1] in Page.TRANSFORM) or
                (ext and ext[1:] in Page.transformers))


class Site(Config):
    """ controls the site and holds the global context object. the context
        object contains all layouts, all posts and categories.
        hooks can be used to configure the context.
    """

    CONTEXT = AttrDict()
    IGNORE = ('_', '.')

    def __init__(self):
        super(Site, self).__init__()

        if not self.LIB_DIR in sys.path and os.path.isdir(self.LIB_DIR):
            sys.path.append(self.LIB_DIR)

        self.layouts = {}

        self.hooks()

        self.context = Site.CONTEXT.copy()
        if not 'site' in self.context:
            self.context.site = AttrDict()

        self.context.site.now = datetime.datetime.now()

    def hooks(self):
        """ load all available hooks from the _hooks/ directory.
        """
        if os.path.isdir(self.HOOK_DIR):
            for f in sorted(self.ignoreFilter(os.listdir(self.HOOK_DIR))):
                if f.endswith('.py'):
                    execfile(os.path.join(self.HOOK_DIR, f), globals())

    def prepare(self):
        """ read all layouts
        """
        self.read_layouts()

    def run(self):
        """ generate the site content to the deploy directory.
        """
        self.write_site_content()

        if options.serve != None:
            try:
                port = int(options.serve)
                site.serve(port)
            except ValueError:
                print 'Invalid Port: %s' % options.serve

    def read_layouts(self):
        if os.path.isdir(self.LAYOUT_DIR):
            self.layouts = [Layout(os.path.join(self.LAYOUT_DIR, f),
                                   self.context)
                                for f in self.ignoreFilter(os.listdir(
                                                            self.LAYOUT_DIR))]
            self.layouts = dict((l.name, l) for l in self.layouts)

    def write_site_content(self):
        """ copy site content to deploy directory.

            ignoring all files and directories, if their filename
            begins with a token defined in IGNORE.

            files with and filename ending with an token defined in
            TRANSFORM are transformed via the Page class. all other
            files are simple copied.
        """

        for root, dirs, files in os.walk(self.BASE_DIR):
            base = root.replace(self.BASE_DIR, '')
            base = base.lstrip(os.path.sep)

            for d in self.ignoreFilter(dirs):
                nd = os.path.join(self.DEPLOY_DIR, base, d)
                if not os.path.isdir(nd):
                    os.makedirs(nd)
            dirs[:] = self.ignoreFilter(dirs)

            for f in self.ignoreFilter(files):
                if Page.transformable(f):
                    Page(os.path.join(root, f),
                         self.layouts,
                         self.context).write()
                else:
                    path = os.path.abspath(root)
                    path = path.replace(os.path.abspath(self.BASE_DIR), '', 1)
                    path = path.lstrip(os.path.sep)
                    path = os.path.join(self.DEPLOY_DIR, path)
                    if not os.path.isdir(path):
                        os.makedirs(path)
                    shutil.copy(os.path.join(root, f), os.path.join(path, f))

    def serve(self, port):
        """ serve the deploy directory with a very simple, cgi
            capable web server on 0.0.0.0:<port>.
        """
        from BaseHTTPServer import HTTPServer
        from CGIHTTPServer import CGIHTTPRequestHandler
        os.chdir(self.DEPLOY_DIR)
        httpd = HTTPServer(('', int(port)), CGIHTTPRequestHandler)
        sa = httpd.socket.getsockname()
        print "Serving HTTP on", sa[0], "port", sa[1], "..."
        try:
            httpd.serve_forever()
        except KeyboardInterrupt:
            pass

    def ignoreFilter(self, seq):
        """ filter out files starting with self.IGNORE tokens
        """

        def ignore_filter(item):
            for ign in self.IGNORE:
                if item.startswith(ign):
                    return False
            return True
        return itertools.ifilter(ignore_filter, seq)

    def setupOptions(self, parser):
        parser.add_option('--serve',
                          action = 'store', dest = 'serve',
                          metavar = 'PORT',
                          help = 'Start web server')

        parser.set_defaults(version = False)
        parser.add_option('-v', '--version',
                          action = 'store_true', dest = 'version',
                          help = 'Output version information and exit')


if __name__ == '__main__':
    DEFAULT_PORT = 8080
    parser = OptionParser(usage = 'syntax: %prog [options] <from> [to]')

    base = deploy_path = None
    args = sys.argv[1:]

    for arg in sys.argv[:0:-1]:
        if not arg.startswith('-'):
            if not deploy_path:
                deploy_path = arg
            elif not base:
                base = arg
        elif arg == '--':
            break

    if not base and deploy_path:
        base = deploy_path
        deploy_path = os.path.join(base, '_deploy')

    if base and os.path.isdir(base):
        Config.updateconfig(base, deploy_path)

    site = Site()

    site.setupOptions(parser)
    (options, args) = parser.parse_args(args)

    if options.version:
        print 'growl version %s - %s (%s)' % (__version__,
                                              __copyright__,
                                              __license__)
        sys.exit(0)

    if not base:
        parser.error('"from" parameter missing!')

    if not os.path.isdir(base):
        print 'error: invalid directory: %s' % base
        sys.exit(2)

    try:
        import markdown
        Config.transformers.setdefault('markdown', markdown.markdown)
    except ImportError:
        pass

    try:
        import textile
        Config.transformers.setdefault('textile', textile.textile)
    except ImportError:
        pass

    try:
        # set jinja2 loader to enable template inheritance
        jinja2_env.loader = jinja2.FileSystemLoader(site.LAYOUT_DIR)
    except NameError:
        pass

    site.options = options

    site.prepare()
    site.run()
